### DISCUSSION TITLE
"AM"/"AppMan" v6.2.1-2 and my push towards creating a community driven repository

### DISCUSSION THREAD

**Discussion by ivan-hc** [ivan-hc/AM#326] [Announcements]
With this new sub-release I am working on pushing towards the use of a database external to the one I have built almost entirely alone in these 3 years.

Tonight I began to combine as many variables as possible to simplify this transition.

I'm happy to provide you with this tool, but I don't want to monopolize the application database. Everyone should be free to upload whatever they want into the database and organize a community repository, regardless of me.

I could not be there at any moment and I would find myself with dozens of requests from applications to load, or that will not be loaded. I want to break this chain and allow anyone to organize a community database.

You can already fork my repository and view all the installation scripts I have published.

If you have any questions, you can contact me on Github and [Reddit](https://www.reddit.com/user/am-ivan/).

Ivan

**Comment by ivan-hc** [comment]
@zen0bit @nazdridoy @DLopezJr what do you think about this?

**Reply by zen0bit** [reply]
I think it's nice.

But most of AppImages I found, was already included... :+1:
Few you add on to my request was added fast enoough

So I don't need to create my database (yet?)

**Reply by nazdridoy** [reply]
@ivan-hc it looks neat. I thought about asking you if you are planning to make a way for a 3rd-party repo. It's a great initiative.

**Reply by nazdridoy** [reply]
I'm pleased to see that our argument and debate about whether to make AM modular or not, and ultimately deciding to make it modular, is paying off. Just look at the repo. AM has come a long way from [v5.8.3](https://github.com/ivan-hc/AM/releases/tag/5.8.3).

**Reply by ivan-hc** [reply]
today I also wrote a guide on how to create and enable your own database https://github.com/ivan-hc/neodb

**Reply by ivan-hc** [reply]
@zen0bit the CLI may remain, but there will always be new applications to add, and even better ones.

Do you know what my best AppImage is? [GIMP](https://github.com/ivan-hc/GIMP-appimage). Like many others, it's just a release that I built to fill the gaps among the apps available in the database... but a couple of days ago I got some wonderful news: the GIMP team has decided to create an official one https://github.com/ivan-hc/GIMP-appimage/issues/21 ! When that happens, I'll be happy to point my scripts towards that official release! This was my goal: to stimulate upstream developers to convince them!

It must be said, however, as I always want to underline... that I could disappear at any moment.

The CLI is a tool that can last forever, it is based on BASH and uses basic Linux terminal commands to work. The other apps, however, change. Always. If my goal of officially adopting them is achieved, I will no longer have any reason to keep all those AppImages, and I will be able to dedicate myself to something else.

As I said, the CLI is something that will always remain, and basically just remove the "sync.am" module and the "`$AMCLIPATH -s`" command from the "update.am" module (each module, if absent, will be downloaded if invoked, it's written in the code)... so third party patches will always be available. I hope they dedicate blogs or communities around "AM"/"AppMan", because together we move forward.

@Samueru-sama , did you see the other day? My github account was suspended for 15 minutes and all my works were gone, and there are a lot of them... how do you think I, who just wanted to make AppImage popular again, felt in those minutes? Very bad! I feared depriving thousands of people of this service.

This is why at least the database needs to be maintained by third parties, @zen0bit . I make too many commits, the activity can be judged illicit by Microsoft, and in an instant, after a commit... I could move away from github, for days, months, years... without knowing that after that commit you will no longer have had a service that you consider valid.

That's why I care so much about making "AM"/"AppMan" popular. A large community with its own domain would prevent you from missing out on this service.

**Reply by Samueru-sama** [reply]
Yeah I noticed the delay. 

I hope you're doing ok. I don't know if you want to talk privately about this?

**Reply by ivan-hc** [reply]
@Samueru-sama don't worry, Github had forced the password renewal of some accounts due to suspected illicit activity... and I fell for it too. Maybe I make too many commits? Long story short, they thought it wasn't me... and they suspended me.

**Comment by ivan-hc** [comment]
If you need to test, you can create a repository containing:
- a directory with the name of your architecture (for example x86_64) containing the installation scripts
- a text file with a list of these applications, named "$arch"-apps (for example, x86_64-apps)

finally, add a file named "neodb" in "$AMPATH", containing the URL to your repositor's items, for example:
```
https://raw.githubusercontent.com/USER/REPO/main
```
so, the new two variables included in APP-MANAGER are these:
```
APPSDB="$(<"$AMPATH/neodb")/$arch"
APPSLISTDB="$(<"$AMPATH/neodb")/$arch-apps"
```
this means that installation scripts will be taken from
```
https://raw.githubusercontent.com/USER/REPO/main/x86_64
```
and the list will be
```
https://raw.githubusercontent.com/USER/REPO/main/x86_64-apps
```
so you're free to manage and improve your own apps database or a community one. You don't need me anymore to upload apps.

**Comment by ivan-hc** [comment]
My idea is to source a script "$AMPATH/neodb" containing the following variables:
```
APPSDB="$AMREPO/programs/$arch"
APPSLISTDB="$AMREPO/programs/$arch-apps"
AMCATALOGUE="https://PORTABLE-LINUX-APPS.GITHUB.IO"
AMCATALOGUEMARKDOWNS="https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/apps"
AMCATALOGUEICONS="https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/icons"
```
for now I only added the default values you can already see in APP-MANAGER

of course to change the source directory of the scripts $APPSBD, the list of the apps $APPSLISTDB, the catalogue $AMCATALOGUE, the pages to show in -a from $AMCATALOGUEMARKDOWNS and the icons database $AMCATALOGUEICONS you hve to change the values above.

This is not ready yet, any improvement is wellcome.

@nazdridoy what would you do?

we need to validate the script, for example, by finding such values in the neodb script, if not valid then we will use the defaults, and then use it as source.

Involved modules are the ones containing these variables.

NOTE, "$AMCATALOGUEICONS" is used by `-i`, module [install.am](https://github.com/ivan-hc/AM/blob/main/modules/install.am), in case the installed app have not an icon in /opt/$APP/icons, if scripts are wrote well, we don't need to invoche that variable

Another solution would be to add contitions in the [database.am](https://github.com/ivan-hc/AM/blob/main/modules/database.am) module, containing the options -a, -l and -q, containing all the other variables... **eccept $APPSDB**, the more important one, used by [install.am](https://github.com/ivan-hc/AM/blob/main/modules/install.am), [download.am](https://github.com/ivan-hc/AM/blob/main/modules/download.am) and [sync.am](https://github.com/ivan-hc/AM/blob/main/modules/sync.am), this is the responsible of the installation scripts.

**Reply by nazdridoy** [reply]
i'm sorry @ivan-hc , i haven't been very active on GitHub recently.
I'm quite busy with my studies. If I had more time, I'd gladly review the current implementation. However, based on the release notes and your explanations, it seems like you're on the right track. Nonetheless, I have some suggestions regarding  third-party repos in general. (Some of these may already be implemented or in your plans.)

1. Allow for multiple custom repos to pull from, similar to arch-core, testing, extra, community, multilib. If no custom repo is defined, the default should be yours.
2. Consider making AMCATALOGUE a soft dependency for custom repos. Not everyone might want to maintain a catalogue site.
3. During downloads (using -d & -i), display and let users choose from which repo the installation scripts are sourced. For example: naz-extra/calibre ==> am -i naz-extra/calibre.

**Reply by ivan-hc** [reply]
Uhm... let me see...

1. both options `-i` and `-d` use this command to check if the script is hosted somewhere
```
if curl --output /dev/null --silent --head --fail $APPSDB/$arg  1>/dev/null; then
```
what you say is to add an extra file as a /etc/apt/sources.list and call urls from that list.

So, we can do 
```
elif for url in /path/to/file/with/multiple/urls; then`
        if curl --output /dev/null --silent --head --fail $url/$arg  1>/dev/null; then
...
```
or something. Like "Archimage" does when checks for packages in regular Arch Linux repositories and AUR.

Not a problem, but I need someone that does this setup, maybe by adding multiple directories in the same repo.

2. If AMCATALOGUE has no value, no URL will be shown in `help`/`-h`, nor in the warning messages. Not a problem, they can do whatever they want. Also, a domain can be github.com, sites.google.com... or anything: the user is totally free to add whatever it wants, also its Name/Nikname.

3. this may be connected to point 1, we can also add a `case` if the same item is in multiple repositories.

So we can merge points 1 and 3 and ignore point 2.

All we need is a function to check the item in multiple urls from a list and then allow the user to choose from where to install7download the app.

**Reply by ivan-hc** [reply]
@nazdridoy it should be the last function before `else`, line 219 of [install.am](https://github.com/ivan-hc/AM/blob/main/modules/install.am) module:
```
elif test -f "$AMPATH/multisources.list"; then
```
to test if a file "multisources.list" exists, it should include a list of URLs, and then it must check each of them to see if "$arg" exsists, throug a `for` cycle.

It the result of 
```
if curl --output /dev/null --silent --head --fail $a_url_in_multisources_list/$arg  1>/dev/null; then
```
gives just one result, then install "$arg" directly, else use `case` the same way we do in [rollback.am](https://github.com/ivan-hc/AM/blob/main/modules/rollback.am)

**Reply by ivan-hc** [reply]
@nazdridoy this is a first draft on how it should appear
```
				elif test -f "$AMPATH/multisources.list"; then
					MULTIREPO=$(tail -n +2 "$AMPATH/multisources.list")
					for anyrepo in $MULTIREPO; do
						if curl --output /dev/null --silent --head --fail $anyrepo/$arg  1>/dev/null; then
							echo "$anyrepo" >> "$AMPATH/.cache/multirepo-args"
						fi
					done
					printf "Select a URL from this menu (read carefully) or press CTRL+C to abort:\n-----------------------------------------------------------------------\n"; sleep 1;
					select d in $(cat "$AMPATH/.cache/multirepo-args"); do test -n "$d" && break; echo ">>> Invalid Selection"; done
					cd $AMPATH/.cache; mkdir -p tmp; cd tmp; wget -q "$d"; cd ..; mv ./tmp/$arg ./$arg; rmdir ./tmp
					_install_arg
					rm -R -f "$AMPATH/.cache/multirepo-args"
				else
```
what we miss is a condition if no arguments then skip, if 1 argument then install, if multiple arguments then choose (the latter is already in place)

**Reply by ivan-hc** [reply]
@nazdridoy I did it https://github.com/ivan-hc/AM/commit/9ebb70a4874879df86239a52f4c52634ac2aa2f6 https://github.com/ivan-hc/AM/commit/a3e1fa57713a76101a535ef81ddc44a7b9518b0e

add a file "$AMPATH/multisources.list" containing urls and try

I've still to correct other things before publishing the new release

@zen0bit @Samueru-sama give it a try (the install.am module is still in "dev")

https://github.com/ivan-hc/AM/blob/dev/modules/install.am

**Reply by ivan-hc** [reply]
@nazdridoy now I have another issue... this time is related to the [database.am](https://github.com/ivan-hc/AM/blob/dev/modules/database.am) module: how can I put lists/marktowns from such new listed repositories?

EDIT: I've done, only your beloved "about" is left xD

**Comment by ivan-hc** [comment]
Hi folks new 6.3 release is out https://github.com/ivan-hc/AM/releases/tag/6.3

No new options but important changes that you should be interested in.

**Comment by ivan-hc** [comment]
UPDATE: new changes for the future 6.4 version... it looks more like Arch Linux's PacMan.

The file "neodb" will not be a script but a configuration file... like this:
```
#STATUS=quiet

[My generic repository name]
Source=https://your-domain.net/apps
List=https://your-domain.net/x86_64-apps
Markdowns=https://your-domain.net/info
```
if you enable the flag "STATUS" you'll not receive notifications about third-party repositories usage

I've completed almost all... only Markdowns is left. Here I need the help of @nazdridoy  that is the "adoptive father" of that option, so him knows it more than me.

**Comment by ivan-hc** [comment]
New release 6.4 is here https://github.com/ivan-hc/AM/releases/tag/6.4

Today the best moment was when I tested the `-l` option and seen the message...
```
YOU HAVE INSTALLED 6 APPLICATIONS OUT OF 7840 AVAILABLE
```
this is because I had not yet filtered the 4 times linked the actual x86_64-apps file containing the list of the apps.

Now I've done... but 7840 can be real to reach, now, if you publish more third-party repos

@zen0bit @Samueru-sama @nazdridoy @DLopezJr check it out!

**Reply by ivan-hc** [reply]
You can test this, if you need:
```
#STATUS=quiet

[DLopezJr]
Source=https://raw.githubusercontent.com/DLopezJr/AM/main/programs/x86_64
List=https://raw.githubusercontent.com/DLopezJr/AM/main/programs/x86_64-apps
Markdowns=https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/apps

[nazdridoy]
Source=https://raw.githubusercontent.com/nazdridoy/AM/main/programs/x86_64
List=https://raw.githubusercontent.com/nazdridoy/AM/main/programs/x86_64-apps
Markdowns=https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/apps

[Samueru-sama]
Source=https://raw.githubusercontent.com/Samueru-sama/AM/main/programs/x86_64
List=https://raw.githubusercontent.com/Samueru-sama/AM/main/programs/x86_64-apps
Markdowns=https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/apps

[oSoWoSo]
Source=https://raw.githubusercontent.com/oSoWoSo/AM/main/programs/x86_64
List=https://raw.githubusercontent.com/oSoWoSo/AM/main/programs/x86_64-apps
Markdowns=https://raw.githubusercontent.com/Portable-Linux-Apps/Portable-Linux-Apps.github.io/main/apps
```

try adding/rename random installation scripts in yout x86_64 directory andlist them in the x86_64-apps file, for test pouposes... or try to `download` or `install` an existing app, if recent, it should not appear in the menu of the older fork you did.

However, Markdowns must be improved, it should work like `-i` and `-d`

**Comment by ivan-hc** [comment]
I was thinking about the `-a`/`about` option and how we can solve with [portable-linux-apps.github.io](https://portable-linux-apps.github.io/) dependence....

...the solution may be rewrite all installation scripts and including the description into them, and through the command:
```
wget -q https://url-to-raw-script-file -O -
```
we could `grep` the part of the file we need to read online.

A community-driven database with the effort of more persons would be the best practice to speedup this task. We need a task force for that. 

I wrote almost all these script in three years, but with the effort of more people we can be faster in build the new database.

And from here we would be able to split scripts into different repositories:
```
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/appimages
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/archimages
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/binaries
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/scripts
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/tars
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/zips
https://raw.githubusercontent.com/coomunitidrivendatabaseName/repositoryName/metascripts
...
```
and so on.

This way, Source= and Markdowns= may be merged together under the same script. One script containing also the info.

This would be a project to be done all together or in shifts.

We also need a different design for the installation scripts.

NOTE, I am only suggesting. Since the tool is able to use third-party repositories, I decided to work for my CLI, but you as a community have to decide for yourself... so "my wishes as the King/Emperor and absolute owner of this repo & Master of puppets" are limited only to this repository, not to external databases. The chains break here. I've been clear?

**Comment by ivan-hc** [comment]
UPDATE: I've partially solved the issue about `-a`/`about` for third-party apps

if no markdown via `curl` but the app is in the list (because lists are merged into one), then generate it using the description from the list, but no URL is available, being a third party app

https://github.com/ivan-hc/AM/commit/de3bd803961898be5fd00348ae4351e46d40160f

following the style of @nazdridoy :wink:

**Comment by ivan-hc** [comment]
Hi folks, I've solved the issue about... `about`!

Please, see https://github.com/ivan-hc/AM/releases/tag/6.4.1 for more details.

Ater this, I think this discussion has reached an end.

Furter implementations are wellcome.
